crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1261,6 +1116,160 @@ impl Target<Cursor> for StoredHistoryObject {
+            if key.checkpoint_viewed_at < stored.checkpoint_sequence_number as u64 {
+                continue;
+            }
+
+            let object =
+                Object::try_from_stored_history_object(stored.clone(), key.checkpoint_viewed_at)?;
+            result.insert(*key, object);
+        }
+
+        Ok(result)
+    }
+}
+
+#[async_trait::async_trait]
+impl Loader<LatestAtKey> for Db {
+    type Value = Object;
+    type Error = Error;
+
+    async fn load(&self, keys: &[LatestAtKey]) -> Result<HashMap<LatestAtKey, Object>, Error> {
+        // Group keys by checkpoint viewed at and parent version -- we'll issue a separate query for
+        // each group.
+        #[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Copy)]
+        struct GroupKey {
+            checkpoint_viewed_at: u64,
+            parent_version: Option<u64>,
+        }
+
+        let mut keys_by_cursor_and_parent_version: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+        for key in keys {
+            let group_key = GroupKey {
+                checkpoint_viewed_at: key.checkpoint_viewed_at,
+                parent_version: key.parent_version,
+            };
+
+            keys_by_cursor_and_parent_version
+                .entry(group_key)
+                .or_default()
+                .insert(key.id);
+        }
+
+        // Issue concurrent reads for each group of keys.
+        let futures = keys_by_cursor_and_parent_version
+            .into_iter()
+            .map(|(group_key, ids)| {
+                self.execute_repeatable(move |conn| {
+                    let Some(range) = AvailableRange::result(conn, group_key.checkpoint_viewed_at)?
+                    else {
+                        return Ok::<Vec<(GroupKey, StoredHistoryObject)>, diesel::result::Error>(
+                            vec![],
+                        );
+                    };
+
+                    let filter = ObjectFilter {
+                        object_ids: Some(ids.iter().cloned().collect()),
+                        ..Default::default()
+                    };
+
+                    // TODO: Implement queries that use a parent version bound using an
+                    // `obj_version` table.
+                    let apply_parent_bound = |q: RawQuery| {
+                        if let Some(parent_version) = group_key.parent_version {
+                            filter!(q, format!("object_version <= {parent_version}"))
+                        } else {
+                            q
+                        }
+                    };
+
+                    Ok(conn
+                        .results(move || {
+                            build_objects_query(
+                                View::Consistent,
+                                range,
+                                &Page::bounded(ids.len() as u64),
+                                |q| apply_parent_bound(filter.apply(q)),
+                                apply_parent_bound,
+                            )
```

Comment:
I think you can just use `View::Historical` here; given the changes in this PR perhaps it's a bit of a misnomer, but since we aren't filtering by `object_version` or some other malleable criteria, and instead solely on `object_id`, we can avoid the bulky `LEFT JOIN`s. Similarly in the paginated objects query, https://github.com/MystenLabs/sui/blob/main/crates/sui-graphql-rpc/src/types/object.rs#L1330, I believe the check should've been instead for whether the filter has anything that isn't empty or `object_ids`.

Alternatively, we can also do a single fetch for all object_ids to `objects_snapshot`, parallel fetches to `objects_history` per checkpoint bound, and then filter in app for the latest per group from either snapshot or history.

Reply:
Furthermore, if we're at a cursor for an object that has a yet newer version, using `View::Consistent` will erroneously filter out the object because a newer version exists, although what we really want is the latest version of the object bounded by the checkpoint

Reply:
**TL;DR: I think `View::Historical` is pretty broken, so I won't use it here, I'll add a TODO in the codebase and create a task for fixing it).**

I see what you're saying (although I had to write out both the historical and consistent queries to double check 😅) -- I think you're right that I can use the `Historical` query kind here, but I'm in two minds about it (see below).

> but since we aren't filtering by object_version or some other malleable criteria,

Note that I **am** filtering by `object_version`, but that is what enables use of the historical mode IIRC.

> I believe the check should've been instead for whether the filter has anything that isn't empty or object_ids.

I don't think this is right.  If you only supply `object_ids`, then you are implicitly filtering the live object set at whatever checkpoint you're querying at, so historical queries do not work, because you need the behaviour of filtering outdated candidates (objects that match the filters but aren't latest in the checkpoint). It works for `object_keys` because you are fixing the version of objects you're querying for.

> using View::Consistent will erroneously filter out the object because a newer version exists

I don't think this will happen, because the query to find newer versions is also bounded by the available range, and by the `apply_parent_bound` modifier that I pass to the last argument here.

Finally while `View::Historical` avoids the left joins:

- It doesn't bound the query on the historical side by the available range (so it's going to look into every partition), while still redundantly querying the snapshot which is a summary of some prefix of object history snapshots.
- In the long run, this loader will be split into two: One that handles object lookups by version, and one that handles latest objects. The former of these will be implemented using the `object_versions` table directly, so should hopefully be nice and cheap, and the latter is implemented most efficiently as a form of consistent lookup (i.e. avoiding looking into all object history partitions).

This actually makes me think that the `Historical` mode might be generally broken... imagine a case where you have an `ObjectFilter` that contains both `object_ids` and `object_keys` -- because it has `object_keys` it will be treated as a historical query, but it also fetching objects by ID, so it's going to fetch them from all partitions in `objects_history` and then take the latest version without applying a bound by available range, which is going to break consistency.